### PR TITLE
db-api-0

### PR DESCRIPTION
**Pull request by raymondlee00**
* remove Firebase functionalities 
* start the connection to proprietary database and API (for proposals, users, and forums)
* update user interface for authentication pages

### CODE REVIEW

File: .flutter-plugins-dependencies [hunk 1]
```diff
@@ -1 +1 @@
-{"_info":"// This is a generated file; do not edit or check into version control.","dependencyGraph":[{"name":"cloud_firestore","dependencies":["firebase_core"]},{"name":"firebase_auth","dependencies":["firebase_core"]},{"name":"firebase_core","dependencies":[]}]}
\ No newline at end of file
+{"_info":"// This is a generated file; do not edit or check into version control.","dependencyGraph":[{"name":"cloud_firestore","dependencies":["firebase_core"]},{"name":"firebase_auth","dependencies":["firebase_core","firebase_auth_web"]},{"name":"firebase_auth_web","dependencies":[]},{"name":"firebase_core","dependencies":["firebase_core_web"]},{"name":"firebase_core_web","dependencies":[]}]}
```
Comment (ericlau00) [inline .flutter-plugins-dependencies]: this file says that we shouldn't check into version control. let's gitignore it next time

File: ios/Flutter/flutter_export_environment.sh [hunk 2]
```diff
@@ -1,10 +1,10 @@
 #!/bin/sh
 # This is a generated file; do not edit or check into version control.
-export "FLUTTER_ROOT=C:\flutter"
-export "FLUTTER_APPLICATION_PATH=C:\Users\shark\Documents\solon-mobile"
+export "FLUTTER_ROOT=C:\Users\Owner\flutter"
```
Comment (ericlau00) [inline ios/Flutter/flutter_export_environment.sh]: we need to gitignore this too

File: lib/API/api_connect.dart [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+
+  static Future<Info> connectRoot() async {
+    final response = await http.get(_url);
+    int status = response.statusCode;
+    return status == 200 ? Info.fromJson(json.decode(response.body)) : throw Exception('data not found');
+  }
+
+  static Future<List<User>> connectUsers() async {
+    final response = await http.get(_url + '/users');
```
Comment (ericlau00) [inline lib/API/api_connect.dart]: dart supports formatted strings with curly brackets; let's stay away from concatenating strings

File: lib/API/proposal.dart [hunk 4]
```diff
@@ -0,0 +1,30 @@
+class Proposal {
+  final int pid;
+  final String title;
+  final DateTime startTime;
+  final DateTime endTime;
```
Comment (ericlau00) [inline lib/API/proposal.dart]: i dont think the json returns a dettime object we probably have to parse first

File: CHANGLOG.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# Changelog
```
Comment (ericlau00) [inline CHANGLOG.md]: make sure to update your changes ray

File: lib/API/api_connect.dart [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+
+  static Future<Info> connectRoot() async {
+    final response = await http.get(_url);
+    int status = response.statusCode;
+    return status == 200 ? Info.fromJson(json.decode(response.body)) : throw Exception('data not found');
+  }
+
+  static Future<List<Proposal>> connectProposals() async {
+    final response = await http.get("${_url}/proposals", headers: {HttpHeaders.authorizationHeader: "onlyweknowthiskey"},);
```
Comment (ericlau00) [inline lib/API/api_connect.dart]: remove braces entirely they arent needed with $ in place apparently

Comment (ericlau00) [inline lib/API/api_connect.dart]: json-ify this and secretize maybe put in gpg

File: lib/API/event.dart [hunk 7]
```diff
@@ -0,0 +1,28 @@
+class Event {
+  final int eid;
+  final String title;
+  final String description;
+  final DateTime date;
+  final DateTime dateCreated;
```
Comment (ericlau00) [inline lib/API/event.dart]: datetime is a string

File: lib/API/info.dart [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  final String info;
+
+  Info({this.info});
+
+  factory Info.fromJson(Map<String, dynamic> json) {
+    return Info(
+      info: json['message']
+    );
+  }
+}
```
Comment (ericlau00) [inline lib/API/info.dart]: this can be deleted once we have something useful to display on the homepage

File: lib/API/secret.dart [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+class SecretLoader {
+
+  static Future<Secret> load(path) {
+    return rootBundle.loadStructuredData<Secret>(path,
+        (jsonStr) async {
+      final secret = Secret.fromJson(json.decode(jsonStr));
+      return secret;
+    });
+  }
+}
```
Comment (ericlau00) [inline lib/API/secret.dart]: we do not need this we can just have the api connect file directly parse json i think

File: lib/API/secrets.json [hunk 10]
```diff
@@ -0,0 +1,3 @@
+{
+    "apikey": "onlywehavethiskey"
+}
```
Comment (ericlau00) [inline lib/API/secrets.json]: blunder pushing this please add to gitignore. thank the lord that we can change headers. also be sure to call this a api_token but key

File: lib/API/info.dart (hunk 8 above)
Comment (ericlau00) [reply lib/API/info.dart]: no longer deleting this because it's useful to get messages

**Comment by ericlau00** [comment]
@raymondlee00 update the CHANGELOG

File: .gitignore [hunk 11]
```diff
@@ -76,4 +79,4 @@ solon-mobile.code-workspace
 !/packages/flutter_tools/test/data/dart_dependencies_test/**/.packages
 
 # Secrets
-android/app/google-services.json
\ No newline at end of file
+android/app/google-services.json
```
Comment (ericlau00) [inline .gitignore]: need to add api token to this and once we completely phase out firebase, we can remove google-services

File: assets/secret [hunk 12]
```diff
@@ -0,0 +1 @@
+onlyweknowthiskey
```
Comment (ericlau00) [inline assets/secret]: @kazijamal lmk when you generate a new header token

File: lib/api/api_connect.dart [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+import 'package:Solon/api/comment.dart';
+import 'package:Solon/api/event.dart';
+import 'package:Solon/api/forumpost.dart';
+import 'package:Solon/api/message.dart';
+import 'package:Solon/api/vote.dart';
+import 'dart:io';
+import 'package:flutter/services.dart';
+import 'package:intl/intl.dart';
+
+class APIConnect {
```
Comment (ericlau00) [inline lib/api/api_connect.dart]: we should consider splitting this file into several depending on the routes

File: lib/api/api_connect.dart [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        ? Message.fromJson(json.decode(response.body)['message'])
+        : throw Exception('data not found');
+  }
+
+  static Future<Vote> connectVotes() async {
+    final response = await http.get(
+      "$_url/votes",
+      headers: {HttpHeaders.authorizationHeader: await loadHeader()},
+    );
+  }
```
Comment (ericlau00) [inline lib/api/api_connect.dart]: this isnt complete yet

File: lib/api/api_connect.dart [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+import 'package:flutter/services.dart';
+import 'package:intl/intl.dart';
+
+class APIConnect {
+  static String _url = "https://api.solonedu.com";
+
+  static Future<String> loadHeader() async {
+    return await rootBundle.loadString('assets/secret');
+  }
+
```
Comment (ericlau00) [inline lib/api/api_connect.dart]: lets consider making this a variable so we don't call loadHeader a bunch of times

File: lib/api/api_connect.dart [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  static Future<Message> connectRoot() async {
+    final response = await http.get(_url);
+    int status = response.statusCode;
+    return status == 200
+        ? Message.fromJson(json.decode(response.body)['message'])
+        : throw Exception('data not found');
```
Comment (ericlau00) [inline lib/api/api_connect.dart]: we can probably have better exception messages.

File: lib/proposal/screen.dart [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+            future: APIConnect.connectProposals(),
+            builder: (context, snapshot) {
+              if (snapshot.hasData) {
+                List<ProposalCard> proposals =
+                    snapshot.data.map((i) => buildProposal(i)).toList();
+                return ListView(
+                  children: proposals,
+                );
+              }
+              return CircularProgressIndicator();
```
Comment (ericlau00) [inline lib/proposal/screen.dart]: why use this instead of loader?

File: lib/api/proposal.dart [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+    var map = new Map<String, dynamic>();
+    map['title'] = title;
+    map['description'] = description;
+    map['starttime'] = startTime;
+    map['endtime'] = endTime;
+    map['uid'] = uid;
+    print("$title $description $startTime $endTime $uid");
+    print("MAP $map");
+    return map;
+  }
```
Comment (ericlau00) [inline lib/api/proposal.dart]: what is the use of this function?

Comment (raymondlee00) [reply lib/api/proposal.dart]: it resturns a JSON based on the proposal attributes that is needed for a POST request. In APIConnect, i take that JSON, encode it as a JSON response and send it out. I made these toMap() functions for proposal, forum post, events, comments, and a bunch others. You can see the article here: https://androidkt.com/http-post-request-in-flutter/ You can refactor those fxn names all to toMap() if you'd like, similar to the fromJSON fxns.

File: lib/proposal/screen.dart (hunk 17 above)
Comment (raymondlee00) [reply lib/proposal/screen.dart]: use loader

File: lib/api/api_connect.dart (hunk 16 above)
Comment (raymondlee00) [reply lib/api/api_connect.dart]: check out my slightly better exceptions msgs

File: lib/api/api_connect.dart (hunk 14 above)
Comment (raymondlee00) [reply lib/api/api_connect.dart]: working on it...im thinking that when i connect to the api and get a votes json response, i check what the message is and based on that, know if that user voted or not.

File: lib/api/api_connect.dart (hunk 13 above)
Comment (raymondlee00) [reply lib/api/api_connect.dart]: maybe in the model class files?

File: lib/api/api_connect.dart [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+
+  //for StreamBuilder
+  // final StreamController<List<Proposal>> _proposals =
+  //     StreamController<List<Proposal>>();
+
+  // final StreamController<List<ForumPost>> _forumposts =
+  //     StreamController<List<ForumPost>>();
+
+  // Stream<List<Proposal>> get proposals => _proposals.stream;
+  // Stream<List<ForumPost>> get forumposts => _forumposts.stream;
```
Comment (ericlau00) [inline lib/api/api_connect.dart]: delete these

File: pubspec.yaml [hunk 20]
```diff
@@ -16,16 +16,28 @@ version: 1.0.0+1
...
+  intl:
+  encrypt:
+  http:
+  asn1lib:
+  pointycastle:
+  password:
+  flutter_string_encryption:
+  rxdart:
+  dio:
+  date_format:
```
Comment (ericlau00) [inline pubspec.yaml]: i dont think we need all of these. some were from encryption

**Pull request merged**